### PR TITLE
Simplify gapfind hole detection

### LINKED ISSUE #768: gapfind produces incorrect set of heights,tasks to fill
Trying to reconcile an internal "lily data completeness" query, I believe I've identified some cases where gapfind does not detect holes and incorrectly produces holes which don't need filling.

Within `chain/find.go` you execute `GapIndexer.findTaskEpochGaps` w the following query (which I've refactored without adjusting the behavior for clarity with comments included inline indicating the logic bugs):

```sql
with
all_heights_and_tasks as (
	select height, task
    from visor_processing_reports
    where height between ? and ?
	group by 1, 2
)
, incomplete_heights as (
	select height, count(task)
	from all_heights_and_tasks
	group by 1
        -- this incorrectly handles NULL_ROUND heights which
        -- only have a single task (gap_find or consensus)
        -- PS: 14 comes from len(AllTasks) in the calling go function
	having count(task) != 14
)
, incomplete_heights_and_their_completed_tasks as (
	select ih.height, pr.task, pr.status_information
	from incomplete_heights ih
	left join visor_processing_reports pr using (height)

	-- this condition excludes NULL_ROUND tasks
	-- which are completed and should be included
	where pr.status_information != 'NULL_ROUND'

	-- what is this filter for?
	-- this will accidentally include rows `ih`
	-- which don't have a joined row on `pr` (which
	-- happens when there's no task history on those heights)
	-- creating heights with no completed tasks at all
	or pr.status_information is null
	group by 1, 2, 3
)
select height, task
from incomplete_heights_and_their_completed_tasks
order by height desc
;
```

### PR DESCRIPTION
**Pull request by placer14**
- removed bugs around edges cases
- this includes the diffing logic within the query which was previously happening after the query in go
- this logic appears to be complete according to spot checks in both directions (ensure this query points at heights that actually contain gaps as well as gaps manually perceived in visor reports which appear in the results of the query)
- gap find no longer writes NULL_ROUNDs to the processing reports table (this separates the concerns of fill and find more clearly, removes special cases for easier querying going forward, and had the side-effect of making test logic simpler)
- this speeds up execution (casual observations suggest around 10x improvement)

fixes #768

### CODE REVIEW

File: chain/find.go [hunk 1]
```diff
@@ -197,90 +199,132 @@ type TaskHeight struct {
...
+
+-- generate all heights in range
+interesting_heights as (
+	select *
+	from generate_series(
+		?0,
+		?1
+	)
+	as x(height)
+)
```
Comment (iand) [inline chain/find.go:226]: :heavy_check_mark:

File: chain/find.go [hunk 2]
```diff
@@ -197,90 +199,132 @@ type TaskHeight struct {
...
+		-- ('actorstatesmultisig'),
+		-- ('actorstatesverifreg'),
+		-- ('blocks'),
+		-- ('messages'),
+		-- ('chaineconomics'),
+		-- ('msapprovals'),
+		-- ('implicitmessage'),
+		-- ('consensus')
+	) as x(task)
+)
```
Comment (iand) [inline chain/find.go:249]: :heavy_check_mark:

File: chain/find.go [hunk 3]
```diff
@@ -197,90 +199,132 @@ type TaskHeight struct {
...
+-- a null round with no data to index.
+-- then take cross product of these heights w tasks
+null_round_heights_and_tasks_in_range as (
+	select pr.height, t.task
+	from visor_processing_reports pr
+	cross join interesting_tasks t
+	where pr.status_information = ?3
+	and pr.height between ?0 and ?1
+	group by 1, 2
+)
```
Comment (iand) [inline chain/find.go:271]: :heavy_check_mark:

File: chain/find.go [hunk 4]
```diff
@@ -197,90 +199,132 @@ type TaskHeight struct {
...
+    select height, task
+    from visor_processing_reports
+    where status = ?3
+		and height between ?0 and ?1
+
+    -- remove the null rounds by height and task
+    except
+    select height, task
+    from null_round_heights_and_tasks_in_range
+)
```
Comment (iand) [inline chain/find.go:293]: :heavy_check_mark:

File: chain/find.go [hunk 5]
```diff
@@ -197,90 +199,132 @@ type TaskHeight struct {
...
+	) as x(task)
+)
+,
+
+-- cross product of heights and tasks
+all_heights_and_tasks_in_range as (
+	select h.height, t.task
+	from interesting_heights h
+	cross join interesting_tasks t
+)
```
Comment (iand) [inline chain/find.go:257]: :heavy_check_mark:

**Review by iand — APPROVED** [review]
LGTM. The :heavy_check_mark:  are just where I was thinking through each stage of the query

**Comment by placer14** [comment]
I've been poking at this query and it seems to be pretty solid for the expected behavior we're looking for.
- I've ensured that there are no rows from the original query (as well as our internal "completeness" report's query) which are missed by the new one
- I've done some spot checking on the false negatives missed by the original query and included in the new one to ensure they are accurate

I've also fixed some tests which were incorrectly ignoring the null round height (which should have 14 missing tasks each being returned by the find function).

**Comment by placer14** [comment]
I found [this bug](https://github.com/filecoin-project/lily/issues/773) and the above query fixes does resolve this case. Will make sure I add a unit test for this. 👏

File: chain/find.go [hunk 6]
```diff
@@ -56,231 +58,158 @@ func (g *GapIndexer) Run(ctx context.Context) error {
...
-	gapReport := make([]*visor.GapReport, 0, len(missingHeights))
-	// walk the possible gaps and query lotus to determine if gap was a null round or missed epoch.
-	for _, gap := range missingHeights {
-		select {
-		case <-ctx.Done():
-			return nil, nil, ctx.Err()
-		default:
-		}
-		gh := abi.ChainEpoch(gap)
-		tsgap, err := node.ChainGetTipSetByHeight(ctx, gh, types.EmptyTSK)
```
Comment (frrist) [inline chain/find.go:171]: This changes the behavior of gap find by causing it to label undetected null rounds as gaps. I think it will be better to address this as a separate change taking into consideration the conversation here: https://github.com/filecoin-project/lily/issues/767

File: chain/find_test.go [hunk 7]
```diff
@@ -8,15 +8,12 @@ import (
 	"testing"
 	"time"
 
-	"github.com/filecoin-project/go-address"
```
Comment (frrist) [inline chain/find_test.go:11]: These changes should be removed and addressed separately in a different PR that fixes https://github.com/filecoin-project/lily/issues/767

**Review by frrist — COMMENTED** [review]
Good stuff. 
I think this PR will be most impactful if it focuses on improving the find query, fixing https://github.com/filecoin-project/lily/issues/773, and adding a unit test for it (see https://github.com/filecoin-project/lily/pull/774).

**Comment by placer14** [comment]
This PR has served it's purpose. Closing and breaking into smaller pieces.
